tutorials/math/matrix-exponentiation.md

```diff
@@ -9,8 +9,522 @@ keywords:
+| 3.    | `fib(n)`   | $n:$ integer                     | Returns $n^{th}$ term of fibonacci series          |
+
+Let us implement them one by one. First one is `multiply(A, B)`.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@Ishwarendra"/>
+
+```cpp
+std::vector<std::vector<int>> multiply(const std::vector<std::vector<int>> A,
+const std::vector<std::vector<int>> B) {
+  int n1 = std::size(A), m1 = std::size(A[0]);
+  int n2 = std::size(B), m2 = std::size(B[0]);
+
+  // If m1 != n2, program will give runtime error
+  assert(m1 == n2); 
+
+  std::vector<std::vector<int>> C(n1, std::vector<int>(m2));
+  for (int i = 0; i < n1; i++) {
+    for (int j = 0; j < m2; j++) {
+      for (int k = 0; k < m1; k++) C[i][j] += A[i][k] * B[k][j];
+    }
+  }
+
+  return C;
+}
+```
+
+</TabItem>
+</Tabs>
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@Ishwarendra"/>
+
+```cpp
+std::vector<std::vector<int>> power(std::vector<std::vector<int>> A, int exp) {
+  int n = std::size(A), m = std::size(A[0]);
+  assert(n == m);
+
+  // We need to start with identity matrix because (A^0 = Identity Matrix)
+  std::vector res(n, std::vector<int>(n));
+  for (int i = 0; i < n; i++) res[i][i] = 1;
+
+  while (exp > 0) {
+    if (exp & 1) res = multiply(res, A);
+
+    A = multiply(A, A);
+    exp >>= 1;
+  }
+  return res;
+}
+```
+
+</TabItem>
+</Tabs>
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@Ishwarendra"/>
+
+```cpp
+int fib(int n) {
+  std::vector<std::vector<int>> T {{1, 1}, {1, 0}};
+  std::vector<std::vector<int>> A {{1}, {0}};
+
+  /*
+    T^n * A = B = {{F(1 + n)},
+                   {F(0 + n)}}
+  */
+  auto B = multiply(power(T, n), A);
+  return B[1][0];
+}
+```
+
+</TabItem>
+</Tabs>
+
+In the given problem it was guaranteed that answer would fit in signed integer range. But if instead of exact number, the problem asked for $fib(n) \mod P,$ where $P = 1e9 + 7$ then we will have to modify our `multiply` function so it performs modulo operation while multiplying the numbers. 
+
+The only change is in the block inside innermost $\textbf{for-loop}$. The following line:
+```cpp
+C[i][j] += A[i][k] * B[k][j];
+```
+
+changes to
+```cpp
+// Multiplying A[i][k] with 1LL converts it into long long type, 
+// which prevents overflow during multiplication. 
+// Max Value of A[i][k] = P - 1 and max value of B[k][j] = P - 1. Both of them fit in integer range
+// But there product doesn't fit in the integer range before we perform modulo operation. 
+C[i][j] += (1LL * A[i][k] * B[k][j]) % P; 
+C[i][j] %= P;
+```
+
+- **Time Complexity:** $O(log(n))$
+- **Space Complexity:** $O(1)$ as we only made matrices of constant dimension.
+
+### [2. Count Vowels Permutation](https://leetcode.com/problems/count-vowels-permutation/description/)
```

Comment:
Change the title to `Example #2: 1220. Count Vowels Permutation`

Reply:
Fixed.